Magpietts 2503 refactor codebook config

rfejgin opened the pull request:
<del>Note: this is an untested draft to demonstrate the idea - please do not merge.<del>

The idea is that rather than defining the number of codebooks, number of codes per codebook and downsampling factor in the Magpie configuration files, we read them from the codec since these parameters must match for the system to work.

That would mean that Magpie would only need to define the number of "special tokens" (like EOS, BOS) to be added after the audio codec tokens positions in the Magpie embedding table. This makes sense since Magpie is the only one that needs to be aware of those.

Ryan L. is adding APIs to the codec so that we can read the relevant parameters; for now we have temporarily implemented this by reaching into the codec (wouldn't work for all codec flavors).

rfejgin committed:
Bugfix: num_audio_tokens_per_codebook

Make sure to reserve enough tokens for special uses like EOS/BOS.

WARNING: old models will be incompatible with the updated inference YAMLs
and will need to override the num_audio_tokens_per_codebook to the value they were
trained with.

rfejgin committed:
Rework how number of codes and codebooks are handled (WIP)

rfejgin commented on nemo/collections/tts/models/magpietts.py:
```diff
@@ -182,12 +200,6 @@ def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
                 temperature=15.0,
             )
 
-        codec_model = AudioCodecModel.restore_from(cfg.get('codecmodel_path'), strict=False)
```
Moved above

rfejgin commented on nemo/collections/tts/models/magpietts.py:
```diff
@@ -97,6 +98,34 @@ def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
+        # TODO @rfejgin: replace these to calls to an API at the codec (not quantizer) level once those are added
+        self.num_audio_codebooks = self._codec_model.vector_quantizer.num_groups
+        num_reserved_audio_tokens = cfg.get('num_reserved_audio_tokens', 8)
+        assert num_reserved_audio_tokens >= SpecialAudioToken.NUM_SPECIAL_TOKENS.value, "Are are not enough reserved entries for all special tokens"
```
This assert should catch it if we try to use more special tokens than will fit in the embedding table (and it would also crash anyway now that we put the special codecs right after the audio codec tokens rather than in reverse from the end of the table).

rfejgin commented on examples/tts/conf/magpietts/magpietts_dc_en.yaml:
```diff
@@ -20,8 +20,7 @@ model:
-  num_audio_codebooks: 8
-  num_audio_tokens_per_codebook: 2018 # 2016 from codec + Keep at least 2 extra for eos/bos ids
+  num_reserved_audio_tokens: 8
   codec_model_downsample_factor: 1024
```
Hoping to also remove the `codec_model_downsample_factor` if we get a suitable API to read it from the codec

rfejgin replied:
This is now read from the codec

rfejgin committed:
Reorder the code a bit for clarity

blisc commented on nemo/collections/tts/models/magpietts.py:
```diff
@@ -165,13 +183,13 @@ def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
                 sa_n_heads=self.cfg.get('local_transformer_n_heads', 1),
                 kernel_size=1,
                 is_causal=True,
-                max_length_causal_mask=cfg.num_audio_codebooks+2,
+                max_length_causal_mask=self.num_audio_codebooks+2,
```
Question for paarth, why is there a +2 here?

blisc replied:
@paarthneekhara

paarthneekhara replied:
It's because the input to the local transformer is BOS + the eight codebooks, and the output is the eight codebooks + EOS. So max_length_causal_mask should be atleast num_audio_codebooks + 1, but I took +2 just to be safe, incase we decide to add any other special token. 

In the ADLR implementation max_length_causal_mask was hardcoded in the transformer class to be something like 4096 (large enough for any sequence.. )

blisc commented on nemo/collections/tts/models/magpietts.py:
```diff
@@ -97,6 +98,34 @@ def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
+        self.num_audio_codebooks = self._codec_model.vector_quantizer.num_groups
+        num_reserved_audio_tokens = cfg.get('num_reserved_audio_tokens', 8)
+        assert num_reserved_audio_tokens >= SpecialAudioToken.NUM_SPECIAL_TOKENS.value, "Are are not enough reserved entries for all special tokens"
+        self.num_audio_tokens_per_codebook = self._codec_model.vector_quantizer.codebook_size_per_group + num_reserved_audio_tokens
+        self.audio_bos_id = self.num_audio_tokens_per_codebook + SpecialAudioToken.AUDIO_BOS.value
+        self.audio_eos_id = self.num_audio_tokens_per_codebook + SpecialAudioToken.AUDIO_EOS.value
+        self.context_audio_bos_id = self.num_audio_tokens_per_codebook + SpecialAudioToken.AUDIO_CONTEXT_BOS.value
+        self.context_audio_eos_id = self.num_audio_tokens_per_codebook + SpecialAudioToken.AUDIO_CONTEXT_EOS.value
```
`self.num_audio_tokens_per_codebook` is a misnomer. It is the total number of tokens in a codebook, maybe we should rename it to `self.num_tokens_per_codebook` or `self.num_all_tokens_per_codebook` or `self.num_total_tokens_per_codebook`.

rfejgin replied:
Agree, will rename

blisc commented on nemo/collections/tts/models/magpietts.py:
```diff
@@ -97,6 +98,34 @@ def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
+        self.num_audio_codebooks = self._codec_model.vector_quantizer.num_groups
+        num_reserved_audio_tokens = cfg.get('num_reserved_audio_tokens', 8)
+        assert num_reserved_audio_tokens >= SpecialAudioToken.NUM_SPECIAL_TOKENS.value, "Are are not enough reserved entries for all special tokens"
+        self.num_audio_tokens_per_codebook = self._codec_model.vector_quantizer.codebook_size_per_group + num_reserved_audio_tokens
+        self.audio_bos_id = self.num_audio_tokens_per_codebook + SpecialAudioToken.AUDIO_BOS.value
+        self.audio_eos_id = self.num_audio_tokens_per_codebook + SpecialAudioToken.AUDIO_EOS.value
+        self.context_audio_bos_id = self.num_audio_tokens_per_codebook + SpecialAudioToken.AUDIO_CONTEXT_BOS.value
+        self.context_audio_eos_id = self.num_audio_tokens_per_codebook + SpecialAudioToken.AUDIO_CONTEXT_EOS.value
```
```suggestion
        self.num_audio_tokens_per_codebook = self._codec_model.vector_quantizer.codebook_size_per_group
        self.audio_bos_id = self.num_audio_tokens_per_codebook + SpecialAudioToken.AUDIO_BOS.value
        self.audio_eos_id = self.num_audio_tokens_per_codebook + SpecialAudioToken.AUDIO_EOS.value
        self.context_audio_bos_id = self.num_audio_tokens_per_codebook + SpecialAudioToken.AUDIO_CONTEXT_BOS.value
        self.context_audio_eos_id = self.num_audio_tokens_per_codebook + SpecialAudioToken.AUDIO_CONTEXT_EOS.value
        self.num_audio_tokens_per_codebook += num_reserved_audio_tokens
```

blisc commented on nemo/collections/tts/models/magpietts.py:
```diff
@@ -97,6 +98,34 @@ def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
+        self._codec_model = codec_model
+        self._codec_model.freeze()  #Lightning does requires_grad = False and self.eval()
+
+        # Reserve special tokens (appended at the end of the codebook after the codec tokens)
+        # (the actual index is this value plus the number of codec tokens - do not use the Enum directy)
+        class SpecialAudioToken(Enum):
+            AUDIO_BOS = 0
+            AUDIO_EOS = 1
+            AUDIO_CONTEXT_BOS = 2
+            AUDIO_CONTEXT_EOS = 3
+            MASK_TOKEN = 4
+            NUM_SPECIAL_TOKENS = 5 # update this if you add more special tokens
```
Should we leave this as the source of truth for the number of special tokens? And delete `num_reserved_audio_tokens` from the configs?

rfejgin replied:
I was thinking the same but was wondering if we want to keep the size of the embedding tables fixed at some value (greater than the number of special tokens + audio tokens) such that it's later possible to load an old checkpoint to new code that supports more special tokens; hence keeping a fixed upper limit on fixed tokens. What do you think?

blisc replied:
Maybe we can add an `override_tokens_per_codebook` param?

blisc commented on nemo/collections/tts/models/magpietts.py:
```diff
@@ -97,6 +98,34 @@ def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
+        self._codec_model = codec_model
+        self._codec_model.freeze()  #Lightning does requires_grad = False and self.eval()
+
+        # Reserve special tokens (appended at the end of the codebook after the codec tokens)
+        # (the actual index is this value plus the number of codec tokens - do not use the Enum directy)
+        class SpecialAudioToken(Enum):
+            AUDIO_BOS = 0
+            AUDIO_EOS = 1
+            AUDIO_CONTEXT_BOS = 2
+            AUDIO_CONTEXT_EOS = 3
+            MASK_TOKEN = 4
+            NUM_SPECIAL_TOKENS = 5 # update this if you add more special tokens
```
We should remove `NUM_SPECIAL_TOKENS` from the Enum and just use `len(SpecialAudioToken)` or other built-in python function instead of defining it.

rfejgin replied:
Indeed, I'll change it. That's the C programmer in me showing up 😂.

blisc commented:
Conversation about old checkpoints:

- Because older checkpoints fixed tokens according to total_tokens and negative indices, however the new codebase does it with positive indices.
- We need to be able to fix the total number of tokens
- Can we enable a way for people to "hardcode" the special tokens? Probably

rfejgin committed:
Merge remote-tracking branch 'blisc/magpietts_2503' into magpietts_2503_refactor_codebook_config

rfejgin committed:
Refactor codebook configuration

* read codec parameters from codec checkpoint; remove corresponding configuration from Magpie YAML files
* add mechanism for backward compatibility with older checkpoints:
** If using `infer_and_evaluate.py`, just set the --legacy_codebooks command line flag
** If running training or inference with the Hydra command line, override using the following flags:
```
forced_num_all_tokens_per_codebook: 2048
forced_audio_bos_id: ${sum:${model.forced_num_all_tokens_per_codebook}, -1}           # 2047
forced_audio_eos_id: ${sum:${model.forced_num_all_tokens_per_codebook}, -2}           # 2046
forced_context_audio_bos_id: ${sum:${model.forced_num_all_tokens_per_codebook}, -4}   # 2044
forced_context_audio_eos_id: ${sum:${model.forced_num_all_tokens_per_codebook}, -3}   # 2045
```

rfejgin committed:
Add README on the codebook reorganization

... and how to load legacy checkpoints.

rfejgin committed:
Cleanup

rfejgin commented:
I have updated the implementation as follows:
* We no longer read any of these from configuration: `codec_model_downsample_factor`, `num_audio_codebooks,  `num_all_tokens_per_codebook`. Instead they are read from the codec and then we add enough indices for all special tokens (plus a few reserved spots for forward-compatibility).
* Added an override mechanism for old checkpoint. It will be activated automatically if `--legacy_codebooks` is passed to `infer_and_evaluate.py`. Alternatively one can manually define overrides like `forced_num_all_tokens_per_codebook` in the YAML file.
* Added a README explaining the changes and how to load old and new checkpoints. See `examples/tts/README_magpietts_legacy_checkpoints.md`.

I believe this addresses all open topics.

rfejgin committed:
Cleanup and fixing typos

rfejgin committed:
Cleanup

rfejgin commented:
hi @blisc and @paarthneekhara I think this is ready now - could you take a look?
For more details see the [README](https://github.com/blisc/NeMo/blob/2edc61a97c4d5f887b8dee11119a307d2c0e8004/examples/tts/README_magpietts_legacy_checkpoints.md) and this [comment](https://github.com/blisc/NeMo/pull/64#issuecomment-2831702132) summarizing the changes. We can also get together to review.

rfejgin committed:
Cleanup

blisc commented on examples/tts/README_magpietts_legacy_checkpoints.md:
````diff
@@ -0,0 +1,77 @@
+With the most common codec setup (2016 codes), the layout used to look like this:
+```
+[0]     Codec Token 0 
+[0]     Codec Token 1
````
Is this supposed to be a [1]?
Can you add some headings to the table to make it easier to understand?

rfejgin replied:
Yes, that's a typo - thanks. Fixed.
I've also updated the table with headings.

rfejgin committed:
Clarify the README on the embedding table layout

rfejgin committed:
README cleanup

shehzeen commented on examples/tts/README_magpietts_legacy_checkpoints.md:
```diff
@@ -0,0 +1,82 @@
+
+## If using a Hydra command line
+You have two options:
+### Add these to your command line
```
This makes sense. We should add a line here saying that we can add these arguments in our training command if we are finetuning (or doing preference alignment) starting from a legacy checkpoint.

rfejgin replied:
Done (in a [subsequent PR](https://github.com/blisc/NeMo/pull/68))

shehzeen commented:
I went through the changes and they look good to me.

I tested the code with the ICML checkpoint and it works fine with `--legacy_codebooks` in the infer_and_evaluate script. Preference alignment/Finetuning should also work by adding `+model. forced_num_all_tokens_per_codebook` etc in the training command. I added a comment about clarifying this in the Readme.

Minor comment about variable names - We have `codec_model_downsample_factor` and `samples_per_frame` referring to the same thing. We can leave it as is since both names seem self explanatory or decide on one of them.

rfejgin committed:
Rename an attritube for clarity

codec_model_downsample_factor --> codec_model_samples_per_frame

rfejgin commented:
> I went through the changes and they look good to me.
> 
> I tested the code with the ICML checkpoint and it works fine with `--legacy_codebooks` in the infer_and_evaluate script. Preference alignment/Finetuning should also work by adding `+model. forced_num_all_tokens_per_codebook` etc in the training command. I added a comment about clarifying this in the Readme.
> 
> Minor comment about variable names - We have `codec_model_downsample_factor` and `samples_per_frame` referring to the same thing. We can leave it as is since both names seem self explanatory or decide on one of them.

Thank you for the review, @shehzeen . I renamed `codec_model_downsample_factor` to `codec_model_samples_per_frame`. Also added a comment in the README as suggested (it's in a subsequent PR).
